src/name/ip_address.rs

```diff
@@ -359,41 +355,44 @@ pub(crate) fn is_valid_ipv4_address(ip_address: untrusted::Input) -> bool {
                     && !ip_address.at_end()
                 {
                     // No octet can start with 0 if a dot does not follow and if we are not at the end.
-                    return false;
+                    return Err(AddrParseError);
                 }
                 if current_size >= current.len() {
                     // More than 3 octets in a triple
-                    return false;
+                    return Err(AddrParseError);
                 }
                 current[current_size] = number - b'0';
                 current_size += 1;
             }
             _ => {
-                return false;
+                return Err(AddrParseError);
             }
         }
         is_first_byte = false;
 
         if ip_address.at_end() {
             if current_size > 0 && radix10_to_octet(&current[..current_size]) > 255 {
                 // No octet can be greater than 255.
-                return false;
+                return Err(AddrParseError);
             }
             break;
         }
     }
-    dot_count == 3
+    if dot_count != 3 {
+        return Err(AddrParseError);
+    }
+    Ok(IpAddrRef::V4(ip_address_, ipv4_octets(ip_address_)?))
```

Comment:
Sorry, no, I don't think this quite makes sense. Now we still have a routine that's parsing the data separately from the validation. The idea is that we accumulate the output as we validate it, so that the output accumulates from the validated data itself. So for example, in the same place where you currently move to the next octet, write into the output array?

Reply:
Addressed in 541bc2b5826354db63292fff15bf7bb3b5a2827a.